### ISSUE TITLE
Upgrade alex and happy

### ISSUE THREAD

**Issue by emwap** [travis-ci/travis-ci#2221]
`alex` and `happy` needs to be upgraded to enable them to work with GHC 7.8

```
$ alex --version
Alex version 3.0.1, (c) 2003 Chris Dornan and Simon Marlow
$ happy --version
Happy Version 1.18.9 Copyright (c) 1993-1996 Andy Gill, Simon Marlow (c) 1997-2005 Simon Marlow
Happy is a Yacc for Haskell, and comes with ABSOLUTELY NO WARRANTY.
This program is free software; you can redistribute it and/or modify
it under the terms given in the file 'LICENSE' distributed with
the Happy sources.
```

`alex` needs to be at or above version `3.1`.
`happy` needs to be at or above version `1.19`.

**Comment by mietek** [comment]
To work around this issue, you can use [Halcyon](https://halcyon.sh/) to handle all Haskell dependencies.

The underlying Cabal issue which makes using _alex,_, _happy,_ and other `build-tools` problematic is https://github.com/haskell/cabal/issues/220.  With Halcyon, you can declare and automatically install `build-tools` by using the [`HALCYON_SANDBOX_EXTRA_APPS`](https://halcyon.sh/reference/#halcyon_sandbox_extra_apps) option.

See https://github.com/thoughtbot/carnival/pull/121 for an example of declaring _alex,_ _happy,_ and _yesod-bin_ as sandbox extra apps.

**Comment by Blaisorblade** [comment]
> Note that with #2690 (new Haskell Platform) we should also get new Alex and Happy. So maybe this will be solved, when the fix is deployed?

The fix to #2690 is deployed, but this isn't fixed — #3197 explains why.

> See thoughtbot/carnival#121 for an example of declaring alex, happy, and yesod-bin as sandbox extra apps, and using Halcyon on TravisCI.

@mietek, I agree that's probably what I should do — I just need some time and a fresh VM to learn halcyon.sh. (But I'll get there!)
